[WebDriver][BiDi] Move event subscription inside the browser.

lauromoura opened the pull request:
#### 763d3d51da9cb47ec02636021caa3536e2b86d9a
<pre>
[WebDriver][BiDi] Move event subscription inside the browser.
<a href="https://bugs.webkit.org/show_bug.cgi?id=295497">https://bugs.webkit.org/show_bug.cgi?id=295497</a>

Reviewed by BJ Burg.

Moving the implementation from Source/WebDriver into the
UIProcess&apos;s Bidi automation code.

Having subscription on the browser will make it easier checking which
events are enabled, especially non-global events, alongside avoiding
excessive message exchange between the browser and driver from the
increasing number of events supported.

For example, as of around 302979@main, without this patch, the Selenium
`api_example_tests.py` alone sends 303 events back to the driver,
despite not enabling any WebDriver-BiDi event. With this patch, all
events are filtered at the browser level.

In this commit, we also replace the Vectors used to track the
subscription fields with HashSets, preparing the way to support
subscriptions by individual browsing contexts. This helped refactor the
unsubscription logic to better match the spec, looping through the
current subscriptions instead of through the individual event names.

This commit also removes the &apos;context&apos; parameter from
session.unsubscribe, as it was removed from the spec, and simplified the
implementation.

Previously, the context and userContexts parameters to session.subscribe
were ignored. Now, we raised NotImplemented errors, as support for them will be
added in follow up commits. Tests expectations were updated accordingly.

We&apos;ll also improve the protocol generator to generate
a helper method that take care of checking whether a given event is
enabled and dispatch it, instead of the current
eventIsEnabled()+dispatch procedure. It can also generate the list of
event names, which is currently manually kept.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::dispatchBidiMessage):
(WebDriver::Session::eventIsEnabled): Deleted.
(WebDriver::Session::subscribeForEvents): Deleted.
(WebDriver::Session::unsubscribeByIDs): Deleted.
(WebDriver::Session::unsubscribeByEventName): Deleted.
* Source/WebDriver/Session.h:
(WebDriver::EventSubscription::isGlobal const): Deleted.
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::bidiSessionSubscribe): Deleted.
(WebDriver::WebDriverService::bidiSessionUnsubscribe): Deleted.
* Source/WebDriver/WebDriverService.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Automation/BidiEventNames.h: Added.
* Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp: Added.
(WebKit::BidiSessionAgent::BidiSessionAgent):
(WebKit::BidiSessionAgent::subscribe):
(WebKit::BidiSessionAgent::unsubscribeByEventName):
(WebKit::BidiSessionAgent::unsubscribe):
(WebKit::BidiSessionAgent::eventIsEnabled):
* Source/WebKit/UIProcess/Automation/BidiSessionAgent.h: Added.
(WebKit::BidiEventSubscription::isGlobal const):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::willShowJavaScriptDialog):
(WebKit::WebAutomationSession::documentLoadedForFrame):
(WebKit::WebAutomationSession::loadCompletedForFrame):
(WebKit::WebAutomationSession::navigationStartedForFrame):
(WebKit::WebAutomationSession::navigationCommittedForFrame):
(WebKit::WebAutomationSession::navigationFailedForFrame):
(WebKit::WebAutomationSession::navigationAbortedForFrame):
(WebKit::WebAutomationSession::fragmentNavigatedForFrame):
(WebKit::WebAutomationSession::contextCreatedForFrame):
(WebKit::WebAutomationSession::recursivelyEmitContextCreatedEvent):
(WebKit::WebAutomationSession::contextDestroyedForPage):
(WebKit::WebAutomationSession::contextDestroyedForFrame):
(WebKit::WebAutomationSession::dismissCurrentJavaScriptDialog):
(WebKit::WebAutomationSession::acceptCurrentJavaScriptDialog):
(WebKit::WebAutomationSession::logEntryAdded):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::eventIsEnabled):
(WebKit::WebDriverBidiProcessor::emitEventIfEnabled):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* WebDriverTests/TestExpectations.json:

Canonical link: <a href="https://commits.webkit.org/310494@main">https://commits.webkit.org/310494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71eed6f5c104564e283e07bb35f27e5cfa248595

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/154011 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/27266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/20426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/162763 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/107477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/27399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/27119 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/119104 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/107477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/156970 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/27399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/138298 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/99804 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/27399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/18426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/10596 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/27399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/16150 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/165236 "Built successfully") | 
| | [  ~~🛠 ios-safer-cpp~~](https://ews-build.webkit.org/#/builders/174/builds/8436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/17751 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/127195 "Passed tests") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/26813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/22460 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/127348 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/34546 "Built successfully and passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/26737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/137944 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/83316 "Built successfully") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/160/builds/26737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/14732 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/26428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/90520 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/26009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/26239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/26081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

lauromoura commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -767,7 +767,8 @@ void WebAutomationSession::willShowJavaScriptDialog(WebPageProxy& page, const St
         // FIXME: propagate the 'userPromptHandler' from session capabilities.
         auto userPromptHandlerType = Inspector::Protocol::BidiSession::UserPromptHandlerType::Accept;
-        m_bidiProcessor->browsingContextDomainNotifier().userPromptOpened(handleForWebPageProxy(page), userPromptType, userPromptHandlerType, message, m_client->defaultTextOfCurrentJavaScriptDialogOnPage(*this, page).value_or(defaultText.value_or(emptyString())));
+        if (m_bidiProcessor->eventIsEnabled("browsingContext.userPromptOpened"_s, { }))
```
FTR, currently, we're checking if the event is enabled at each point where we might trigger it, in order to avoid creating the event payload and discarding it if we did some "lazy" subscription check instead.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,47 @@
+            "async": true,
+            "parameters": [
+                { "name": "events", "type": "array", "items": { "type": "string" }, "description": "The list of events to subscribe to." },
+                { "name": "contexts", "type": "array", "items": { "$ref": "BidiBrowsingContext.Info" }, "description": "Browsing contexts to subscribe to", "optional": true },
```
The type ref for the array elements should be BidiBrowsingContext (context ID), not the associated Info type.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,47 @@
     "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/session",
     "types": [
+        {
+            "id": "Subscription",
```
Nit: SubscriptionID

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,47 @@
+            "parameters": [
+                { "name": "events", "type": "array", "items": { "type": "string" }, "description": "The list of events to subscribe to." },
+                { "name": "contexts", "type": "array", "items": { "$ref": "BidiBrowsingContext.Info" }, "description": "Browsing contexts to subscribe to", "optional": true },
+                { "name": "userContexts", "type": "array", "items": { "$ref": "BidiBrowser.UserContext" }, "description": "User contexts to subscribe to", "optional": true }
```
If not yet implemented, can you add a link to the bugzilla in the description?

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,47 @@
+            "async": true,
+            "parameters": [
+                { "name": "subscriptions", "type": "array", "items": { "$ref": "Subscription" }, "description": "A unique identifier for the subscription.", "optional": true },
+                { "name": "events", "type": "array", "items": { "type": "string" }, "description": "The list of events to subscribe to.", "optional": true },
```
It would be good to have an enum in the protocol with a list of the supported events. Then this inner type can be BidiSession.EventName

lauromoura replied:
I tried adding it, but got some errors due to the '.' in the name of event, and collision with the module declaration (e.g. `"log"` is a valid parameter here).

What if we add support for this tighter event name type in the context of https://webkit.org/b/291371 (support subscribing to all events in a module if given only the module name)? We could even generate the valid event names from the declared events in the other Bidi json protocol modules, to avoid keeping a manual list.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,47 @@
+            "parameters": [
+                { "name": "subscriptions", "type": "array", "items": { "$ref": "Subscription" }, "description": "A unique identifier for the subscription.", "optional": true },
+                { "name": "events", "type": "array", "items": { "type": "string" }, "description": "The list of events to subscribe to.", "optional": true },
+                { "name": "contexts", "type": "array", "items": { "$ref": "BidiBrowsingContext.Info" }, "description": "Browsing contexts to subscribe to", "optional": true }
```
Ditto comment about passing context ID vs info struct. If not yet implemented please note in the description, using short bugzilla links (webkit.org/b/NNNNNN).

Alternatively, the bug link can be stuffed into an unused property in the spec json, this is the approach taken for each command. But we could have per-parameter links.

What do you think?

lauromoura replied:
I'm going to add it inline into the description for now.

(In fact, I already have a patch for the contexts one (not userContext) as a follow-up to this patch)

burg commented on Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp:
```diff
@@ -0,0 +1,240 @@
+        contexts ? contexts->toJSONString().utf8().data() : "null");
+
+    if (!events || !events->length()) {
+        callback(makeUnexpected(STRING_FOR_PREDEFINED_ERROR_NAME_AND_DETAILS(InvalidParameter, "At least one event name must be provided."_s)));
```
We can probably make use of WebAutomationSessionMacros.h to make these early exit cases shorter.

burg commented on Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp:
```diff
@@ -0,0 +1,240 @@
+        const auto& subscription = m_eventSubscriptions.get(id);
+
+        for (const auto& event : subscription.events) {
+            auto removeResult = m_eventSubscriptionCounts.find(event);
```
Should this be findResult?

burg commented on Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp:
```diff
@@ -0,0 +1,240 @@
+bool BidiSessionAgent::eventIsEnabled(const String& eventName, const HashSet<String>& browsingContexts)
+{
+    // https://w3c.github.io/webdriver-bidi/#event-is-enabled
+
```
Nit: extra newlines

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -1057,27 +1060,32 @@ void WebAutomationSession::didCreatePage(WebPageProxy& page)
 void WebAutomationSession::navigationStartedForFrame(const WebFrameProxy& frame, std::optional<WebCore::NavigationIdentifier> navigationID)
 {
-    m_bidiProcessor->browsingContextDomainNotifier().navigationStarted(effectiveHandleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), WallTime::now().secondsSinceEpoch().milliseconds(), frame.url().string());
+    if (m_bidiProcessor->eventIsEnabled("browsingContext.navigationStarted"_s, { }))
```
Can we extract all the event names, at least into static variables at the top?

lauromoura replied:
Extracted into a (for now) manual header. In the follow up, when we touch the generator, we can make it generated.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -1144,7 +1152,8 @@ void WebAutomationSession::contextCreatedForFrame(const WebFrameProxy& frame)
     }
 
-    m_bidiProcessor->browsingContextDomainNotifier().contextCreated(contextHandle, url, "null"_s, parentHandle, JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create(), clientWindow, userContext);
+    if (m_bidiProcessor->eventIsEnabled("browsingContext.contextCreated"_s, { }))
```
Would prefer that the call be something like

`m_bidiProcessor->sendEventIfEnabled(EventNames::BidiBrowsingContext::ContextCreated, args...)`

That way the calling code doesn't need to keep the event names in sync.

If we want to avoid materializing/marshalling the event arguments needlessly, it could take a lambda that is conditionally called back. But for these simple typed arguments, I think it's fine to pass refs.

lauromoura replied:
Makes sense. IIUC, we'd need one overload of these helpers per event, right? (So it glues the event name to the domain notifier method)

Should I add a manual version of such wrapper to this patch (for the existing events) for validation, and in that follow up I mentioned we actually generate them?

burg reviewed (commented):
I think this is nearly ready to merge, but needs rebase again. Sorry about the inconsistent review pace.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,47 @@
+            "name": "unsubscribe",
+            "description": "Unsubscribes from events from the remote end.",
+            "spec": "https://w3c.github.io/webdriver-bidi/#command-session-unsubscribe",
+            "wpt": "https://github.com/web-platform-tests/wpt/tree/master/webdriver/tests/bidi/session/subscribe/",
```
Nit: wpt link is for `subscribe` not `unsubscribe`

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,46 @@
+            "async": true,
+            "parameters": [
+                { "name": "subscriptions", "type": "array", "items": { "$ref": "SubscriptionID" }, "description": "A unique identifier for the subscription.", "optional": true },
+                { "name": "events", "type": "array", "items": { "type": "string" }, "description": "The list of events to unsubscribe from.", "optional": true }
```
Might want to specify in the description whether this is an array of qualified Domain.eventNames or something else?

burg commented on Source/WebKit/UIProcess/Automation/BidiEventNames.h:
```diff
@@ -0,0 +1,54 @@
+#include <wtf/text/ASCIILiteral.h>
+
+namespace WebKit {
+
```
Nit: probably don't need extra newlines between `namespace {` lines

burg replied:
Nit: probably don't need extra newlines between `namespace {` lines

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
```diff
@@ -6,12 +6,46 @@
+            "async": true,
+            "parameters": [
+                { "name": "events", "type": "array", "items": { "type": "string" }, "description": "The list of events to subscribe to." },
+                { "name": "contexts", "type": "array", "items": { "$ref": "BidiBrowsingContext.BrowsingContext" }, "description": "Browsing contexts to subscribe to. https://webkit.org/b/282981", "optional": true },
```
Should this say "unimplemented" ?

burg commented on Source/WebKit/UIProcess/Automation/BidiEventNames.h:
```diff
@@ -0,0 +1,54 @@
+
+namespace BidiEventNames {
+
+namespace BrowsingContext {
```
Nice.

burg commented on Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp:
```diff
@@ -0,0 +1,210 @@
+
+    HashSet<AtomString> atomEventNames;
+    for (const auto& event : events.get()) {
+        auto eventName = event->asString();
```
Since events and contexts are untyped JSON arrays, we need to bail here if eventName is nullptr or the JSON::Value subtype check  fails.

lauromoura replied:
Isn't this covered by the `ASYNC_FAIL...(name.isEmpty())` check?

burg commented on Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp:
```diff
@@ -0,0 +1,210 @@
+        atomEventNames.add(AtomString { eventName });
+    }
+
+    ASYNC_FAIL_WITH_PREDEFINED_ERROR_AND_DETAILS_IF(contexts && userContexts && (contexts->length() > 0) && (userContexts->length() > 0), InvalidParameter, "Contexts and user contexts cannot be used together."_s);
```
Please write this differently, one case per parameter.

lauromoura replied:
I'm not sure if this is possible, as this check is about the presence of both `contexts`/`userContexts` at the same time.

burg commented on Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp:
```diff
@@ -0,0 +1,210 @@
+    // FIXME: Support by-context subscriptions.
+    // https://bugs.webkit.org/show_bug.cgi?id=282981
+    // Also: spec says we need to convert into top-level browsing contexts.
+    if (contexts && contexts->length())
```
I believe this should return `NotImplemented` if we get to this point.

burg commented on Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp:
```diff
@@ -0,0 +1,210 @@
+    if (contexts && contexts->length())
+        LOG(Automation, "BidiSessionAgent::subscribe: Ignoring contexts parameter, as by-context subscriptions are not supported yet.");
+
+    if (userContexts && userContexts->length())
```
Ditto.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -1072,27 +1076,32 @@ void WebAutomationSession::didCreatePage(WebPageProxy& page)
 void WebAutomationSession::navigationStartedForFrame(const WebFrameProxy& frame, std::optional<WebCore::NavigationIdentifier> navigationID)
 {
-    m_bidiProcessor->browsingContextDomainNotifier().navigationStarted(effectiveHandleForWebFrameProxy(frame), navigationIDToProtocolString(navigationID), WallTime::now().secondsSinceEpoch().milliseconds(), frame.url().string());
+    if (m_bidiProcessor->eventIsEnabled(BidiEventNames::BrowsingContext::NavigationStarted, { }))
```
I'm a little concerned about having to repeat this pattern anywhere an event gets fired. Also this makes it tricky to debug log events that weren't subscribed to. Can it be a lambda?

`m_bidiProcessor->emitEventIfEnabled(eventName, [] { bidiProcessor->browsingContextDomainNotifier().navigationStarted(...)}) })`

lauromoura replied:
I'll replace with the lambdas, which will already might be a step towards the generated handlers I mentioned above.

burg reviewed (approved):
Good work, mostly code style and maintenance tweaks left.

burg reviewed (approved):
Good work, mostly code style and maintenance tweaks left.

lauromoura committed:
[WebDriver][BiDi] Move event subscription inside the browser.
https://bugs.webkit.org/show_bug.cgi?id=295497

Reviewed by BJ Burg.

Moving the implementation from Source/WebDriver into the
UIProcess's Bidi automation code.

Having subscription on the browser will make it easier checking which
events are enabled, especially non-global events, alongside avoiding
excessive message exchange between the browser and driver from the
increasing number of events supported.

For example, as of around 302979@main, without this patch, the Selenium
`api_example_tests.py` alone sends 303 events back to the driver,
despite not enabling any WebDriver-BiDi event. With this patch, all
events are filtered at the browser level.

In this commit, we also replace the Vectors used to track the
subscription fields with HashSets, preparing the way to support
subscriptions by individual browsing contexts. This helped refactor the
unsubscription logic to better match the spec, looping through the
current subscriptions instead of through the individual event names.

This commit also removes the 'context' parameter from
session.unsubscribe, as it was removed from the spec, and simplified the
implementation.

Previously, the context and userContexts parameters to session.subscribe
were ignored. Now, we raised NotImplemented errors, as support for them will be
added in follow up commits. Tests expectations were updated accordingly.

We'll also improve the protocol generator to generate
a helper method that take care of checking whether a given event is
enabled and dispatch it, instead of the current
eventIsEnabled()+dispatch procedure. It can also generate the list of
event names, which is currently manually kept.

* Source/WebDriver/Session.cpp:
(WebDriver::Session::dispatchBidiMessage):
(WebDriver::Session::eventIsEnabled): Deleted.
(WebDriver::Session::subscribeForEvents): Deleted.
(WebDriver::Session::unsubscribeByIDs): Deleted.
(WebDriver::Session::unsubscribeByEventName): Deleted.
* Source/WebDriver/Session.h:
(WebDriver::EventSubscription::isGlobal const): Deleted.
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::bidiSessionSubscribe): Deleted.
(WebDriver::WebDriverService::bidiSessionUnsubscribe): Deleted.
* Source/WebDriver/WebDriverService.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Automation/BidiEventNames.h: Added.
* Source/WebKit/UIProcess/Automation/BidiSessionAgent.cpp: Added.
(WebKit::BidiSessionAgent::BidiSessionAgent):
(WebKit::BidiSessionAgent::subscribe):
(WebKit::BidiSessionAgent::unsubscribeByEventName):
(WebKit::BidiSessionAgent::unsubscribe):
(WebKit::BidiSessionAgent::eventIsEnabled):
* Source/WebKit/UIProcess/Automation/BidiSessionAgent.h: Added.
(WebKit::BidiEventSubscription::isGlobal const):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::willShowJavaScriptDialog):
(WebKit::WebAutomationSession::documentLoadedForFrame):
(WebKit::WebAutomationSession::loadCompletedForFrame):
(WebKit::WebAutomationSession::navigationStartedForFrame):
(WebKit::WebAutomationSession::navigationCommittedForFrame):
(WebKit::WebAutomationSession::navigationFailedForFrame):
(WebKit::WebAutomationSession::navigationAbortedForFrame):
(WebKit::WebAutomationSession::fragmentNavigatedForFrame):
(WebKit::WebAutomationSession::contextCreatedForFrame):
(WebKit::WebAutomationSession::recursivelyEmitContextCreatedEvent):
(WebKit::WebAutomationSession::contextDestroyedForPage):
(WebKit::WebAutomationSession::contextDestroyedForFrame):
(WebKit::WebAutomationSession::dismissCurrentJavaScriptDialog):
(WebKit::WebAutomationSession::acceptCurrentJavaScriptDialog):
(WebKit::WebAutomationSession::logEntryAdded):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
(WebKit::WebDriverBidiProcessor::eventIsEnabled):
(WebKit::WebDriverBidiProcessor::emitEventIfEnabled):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiSession.json:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* WebDriverTests/TestExpectations.json:

Canonical link: https://commits.webkit.org/310494@main
